libnopegl/src/text_external.c

```diff
@@ -0,0 +1,747 @@
+    }
+
+    return 0;
+}
+
+static size_t find_face_with_codepoint(const struct darray *ft_faces_array, FriBidiChar ch, size_t prev_face_id)
+{
+    const FT_ULong charcode = (FT_ULong)ch;
+    const FT_Face *ft_faces = ngli_darray_data(ft_faces_array);
+    for (size_t face_id = 0; face_id < ngli_darray_count(ft_faces_array); face_id++)
+        if (FT_Get_Char_Index(ft_faces[face_id], charcode))
+            return face_id;
+    return SIZE_MAX;
+}
+
+/*
+ * Split the sequence of codepoints into multiple runs (or just one if there is
+ * only one font face needed)
+ */
+static int split_into_runs(struct text *text, struct darray *runs_array,
+                           const FriBidiChar *str, size_t len,
+                           enum run_type type, size_t start, size_t end)
+{
+    struct text_external *s = text->priv_data;
+
+    ngli_assert(start < end);
+
+    size_t prev_face_id = find_face_with_codepoint(&s->ft_faces, str[start], 0);
+    size_t pos = start;
+
+    for (;;) {
+        size_t face_id;
+        do {
+            face_id = find_face_with_codepoint(&s->ft_faces, str[pos], prev_face_id);
+            if (face_id != prev_face_id)
+                break;
+            pos++;
+        } while (pos < end);
+
+        int ret = append_run(text, runs_array, str, len, prev_face_id, type, start, pos);
+        if (ret < 0)
+            return ret;
+
+        if (pos == end)
+            break;
+
+        start = pos;
+        prev_face_id = face_id;
+    }
+
+    return 0;
+}
+
+static int handle_words_and_wordseps(struct text *text, struct darray *runs_array, const FriBidiChar *str, size_t len)
+{
+    size_t pos = 0;
+
+    for (;;) {
+        /* Handle word separators (spaces) */
+        size_t end = pos;
+        while (end < len && char_is_whitespace(str[end]))
+            end++;
+        if (end > pos) {
+            int ret = split_into_runs(text, runs_array, str, len, RUN_TYPE_WORDSEP, pos, end);
+            if (ret < 0)
+                return ret;
+        }
+        if (end == len)
+            break;
+        pos = end;
+
+        /* Hande words */
+        while (end < len && !char_is_whitespace(str[end]))
+            end++;
+        if (end > pos) {
+            int ret = split_into_runs(text, runs_array, str, len, RUN_TYPE_WORD, pos, end);
+            if (ret < 0)
+                return ret;
+        }
+        if (end == len)
+            break;
+        pos = end;
+    }
+
+    return 0;
+}
+
+static int handle_line_breaks(struct text *text, struct darray *runs_array,
+                              const FriBidiChar *str, size_t len, size_t *pos)
+{
+    size_t end = *pos;
+    while (end < len && char_is_linebreak(str[end]))
+        end++;
+    if (end > *pos) {
+        int ret = split_into_runs(text, runs_array, str, len, RUN_TYPE_LINEBREAK, *pos, end);
+        if (ret < 0)
+            return ret;
+        *pos = end;
+    }
+    return 0;
+}
+
+#define STACK_LIST_SIZE 128
```

Comment:
I guess 128 handle most of the cases ? does that correspond to a line/paragraph of 128 chars ?

Reply:
Yup that's exactly that. Incidentally, that's also the value used by fribidi in its `log2vis`

Reply:
Ok. Thanks.